event-exporter/kubernetes/podlabels/cache_collector.go

```diff
@@ -0,0 +1,159 @@
+	emptyLabelPodCache    *lru.Cache[cacheKey, time.Time] // cache the timestamp so we can check TTL.
+	emptyLabelPodCacheTTL time.Duration
+}
+
+// NewCollector returns a new podLabelCollectorWithCache that maintains a set of (podName, namespace) -> (podLabels) mappings.
+func NewCollector(client kubernetes.Interface, ignoredNamespaces []string, podCacheSize, emptyLabelPodCacheSize int, emptyLabelPodCacheTTL, getPodTimeout time.Duration) (*podLabelCollectorWithCache, error) {
+	cache, err := lru.NewWithEvict[cacheKey, map[string]string](podCacheSize, recordEviction)
+	if err != nil {
+		return nil, err
+	}
+	emptyLabelPodCache, err := lru.NewWithEvict[cacheKey, time.Time](emptyLabelPodCacheSize, recordEmptyLabelPodCacheEvict)
+	if err != nil {
+		return nil, err
+	}
+	pc := &podLabelCollectorWithCache{
+		client:                client,
+		getPodTimeout:         getPodTimeout,
+		ignoredNamespaces:     make(map[string]struct{}),
+		cache:                 cache,
+		emptyLabelPodCache:    emptyLabelPodCache,
+		emptyLabelPodCacheTTL: emptyLabelPodCacheTTL,
+	}
+	for _, ns := range ignoredNamespaces {
+		pc.ignoredNamespaces[ns] = struct{}{}
+	}
+
+	return pc, nil
+}
+
+// Get returns the label mappings for a given Pod. If there's no local cache, we will attempt a
+// get from apiserver.
+func (pc *podLabelCollectorWithCache) GetLabels(namespaceName, podName string) map[string]string {
+	if _, ok := pc.ignoredNamespaces[namespaceName]; ok {
+		return nil
+	}
+	if labels, ok := pc.cache.Get(cacheKey{Namespace: namespaceName, Name: podName}); ok {
+		recordQueryHit()
+		return labels
+	}
+	recordQueryMiss()
+	// For pod with empty labels, return empty label directly.
+	if timestamp, ok := pc.emptyLabelPodCache.Get(cacheKey{Namespace: namespaceName, Name: podName}); ok {
+		if timestamp.Add(pc.emptyLabelPodCacheTTL).After(time.Now()) {
+			recordEmptyLabelPodCacheHit()
+			return nil
+		} else {
+			recordEmptyLabelPodCacheExpire()
+			pc.emptyLabelPodCache.Remove(cacheKey{Namespace: namespaceName, Name: podName})
+		}
+	} else {
+		recordEmptyLabelPodCacheMiss()
+	}
+	ctx, cancel := context.WithTimeout(context.TODO(), pc.getPodTimeout)
+	pod, err := pc.client.CoreV1().Pods(namespaceName).Get(ctx, podName, metav1.GetOptions{})
+	cancel()
+	if err != nil {
+		if getPodErrLogRateLimiter.Allow() {
+			glog.Errorf("Failed to get pod %s/%s %v", namespaceName, podName, err)
+		}
+		if statusError, isStatus := err.(*apierrors.StatusError); isStatus {
+			recordPodGet(string(statusError.ErrStatus.Reason))
+			if apierrors.IsNotFound(statusError) {
+				recordEmptyLabelPodCacheAddition()
+				pc.emptyLabelPodCache.Add(cacheKey{Namespace: namespaceName, Name: podName}, time.Now())
+			}
+		} else {
+			recordPodGet("UnknownFailure")
+		}
+		return nil
+	}
+	recordPodGet("OK")
+	labels := getLabelsFromPod(pod)
+	if len(labels) > 0 {
+		pc.cache.Add(cacheKey{Namespace: namespaceName, Name: podName}, labels)
```

Comment:
Should we call pc.emptyLabelPodCache.Add(cacheKey{Namespace: namespaceName, Name: podName}, time.Now()) to refresh TTL here?